File: src/sc_neurocore/studio/evidence_seal.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SC-NeuroCore — Studio cross-runtime evidence seal

"""One digest that means the same thing on the server and in the browser.

Every Studio evidence lane sealed its payloads with ``json.dumps(...,
sort_keys=True)`` and the digest of that text. That text is Python's, not
JSON's. Measured on the real surface before this module existed: a model run of
``AdExNeuron`` recorded ``result_sha256``
``771d8e51…``; the identical payload, after the round trip every exported
artefact makes through the operator's browser, re-sealed to ``813d3005…``.
The values were unchanged — Python compared the two payloads equal — but
``1.0`` had come back as ``1``, ``-70.0`` as ``-70`` and ``1e-07`` as
``1e-7``. No verifier could exist: checking a recorded seal against a
browser-supplied payload would have failed for every honest run.

The seal here is defined over *values*, not over one runtime's rendering of
them. A number is written in one normal form that both runtimes reach from the
same double, keys are ordered by code point rather than by UTF-16 unit, strings
carry the minimal RFC 8259 escapes, and anything that cannot survive the trip
intact — a non-finite float, an integer no double holds exactly, an unpaired
surrogate — is refused instead of silently altered.

``studio/frontend/src/evidenceSeal.ts`` is the byte-identical counterpart.
Neither runtime can perform the other's JSON round trip, so parity is carried
by two committed corpora that both read —
``studio/frontend/src/evidenceSealVectors.json`` (the cases a person would
write down) and ``evidenceSealRandomVectors.json`` (1024 doubles drawn as bit
patterns, written by ``tools/generate_evidence_seal_vectors.py``). Each side
parses the same JSON with its own parser and must reach the same canonical
text. ``tests/test_studio_evidence_seal.py`` holds this side and the freshness
of the corpora; ``evidenceSeal.test.ts`` holds the other and additionally puts
every vector through a real ``JSON.parse(JSON.stringify(...))`` before sealing
it, which is the browser round trip itself.
"""

from __future__ import annotations

from collections.abc import Mapping, Sequence
import hashlib
import math

#: Contract version of the canonical form. A change to the normal form is a
#: change of digest for the same value, so it takes a new version.
EVIDENCE_SEAL_SCHEMA_VERSION = "studio.evidence-seal.v1"

#: Digest algorithm named in every receipt that carries a seal.
EVIDENCE_SEAL_ALGORITHM = "sha256"


_SHORT_ESCAPES = {
    "\b": "\\b",
    "\t": "\\t",
    "\n": "\\n",
    "\f": "\\f",
    "\r": "\\r",
    '"': '\\"',
    "\\": "\\\\",
}


class EvidenceSealError(ValueError):
    """Raised when a value cannot be sealed identically in both runtimes."""


def seal_sha256(value: object) -> str:
    """Return the SHA-256 digest of the canonical form of ``value``.

    Parameters
    ----------
    value : object
        Any JSON-shaped value: ``None``, ``bool``, ``int``, ``float``, ``str``,
        a mapping with string keys, or a sequence of the same.

    Returns
    -------
    str
        Lowercase 64-character hexadecimal digest.

    Raises
    ------
    EvidenceSealError
        If the value contains anything that would not survive a JSON round
        trip through the browser unchanged.
    """
    return hashlib.sha256(canonical_seal_text(value).encode("utf-8")).hexdigest()


def canonical_seal_text(value: object) -> str:
    """Return the canonical JSON text used as the seal input.

    Parameters
    ----------
    value : object
        The value to encode.

    Returns
    -------
    str
        Canonical JSON: object keys in code-point order, no insignificant
        whitespace, numbers in the shared normal form, minimal string escapes.

    Raises
    ------
    EvidenceSealError
        If the value is not JSON-shaped, carries a non-finite float, an integer
        no double holds exactly, or an unpaired surrogate.
    """
    parts: list[str] = []
    _encode(value, parts)
    return "".join(parts)


def _encode(value: object, parts: list[str]) -> None:
    if value is None:
        parts.append("null")
        return
    if isinstance(value, bool):
        parts.append("true" if value else "false")
        return
    if isinstance(value, int):
        parts.append(_canonical_integer(value))
        return
    if isinstance(value, float):
        parts.append(_canonical_float(value))
        return
    if isinstance(value, str):
        parts.append(_canonical_string(value))
        return
    if isinstance(value, Mapping):
        _encode_mapping(value, parts)
        return
    if isinstance(value, Sequence) and not isinstance(value, (bytes, bytearray)):
        _encode_sequence(value, parts)
        return
    raise EvidenceSealError(f"{type(value).__name__} is not a sealable JSON value.")


def _encode_mapping(value: Mapping[object, object], parts: list[str]) -> None:
    keys: list[str] = []
    for key in value:
        if not isinstance(key, str):
            raise EvidenceSealError("A sealable object key must be a string.")
        keys.append(key)
    parts.append("{")
    for index, key in enumerate(sorted(keys, key=_code_points)):
        if index:
            parts.append(",")
        parts.append(_canonical_string(key))
        parts.append(":")
        _encode(value[key], parts)
    parts.append("}")


def _encode_sequence(value: Sequence[object], parts: list[str]) -> None:
    parts.append("[")
    for index, item in enumerate(value):
        if index:
            parts.append(",")
        _encode(item, parts)
    parts.append("]")


def _code_points(value: str) -> tuple[int, ...]:
    """Return the code points of ``value`` for runtime-independent ordering.

    Python orders strings by code point and JavaScript by UTF-16 code unit; the
    two disagree above the basic multilingual plane. Ordering by this key makes
    both runtimes agree.
    """
    return tuple(ord(character) for character in value)


def _canonical_integer(value: int) -> str:
    """Return the token for an integer, refusing one no double holds exactly.

    A JSON number reaches the browser as a double and comes back as one. An
    integer the conversion changes would be sealed under a value it does not
    keep, so it is refused; the magnitude alone does not decide this, because a
    large integer with enough factors of two survives intact.
    """
    try:
        exact = int(float(value)) == value
    except OverflowError:
        exact = False
    if not exact:
        raise EvidenceSealError(
            f"Integer {value} is not carried exactly by a JSON number; "
            "seal it as a string if it must be preserved."
        )
    return str(value)


def _canonical_float(value: float) -> str:
    """Return the normal form of a finite double.

    An integral double is written as the integer it equals, because that is
    what the browser sends back for it. Any other double is written in
    scientific form built from its shortest round-tripping digits, which both
    runtimes produce identically from the same double.
    """
    if not math.isfinite(value):
        raise EvidenceSealError(f"{value!r} is not a sealable JSON number.")
    if value.is_integer():
        return str(int(value))
    significand, _, exponent_text = repr(abs(value)).partition("e")
    integer_text, _, fraction = significand.partition(".")
    # The shortest round-tripping form never ends in a redundant zero, in either
    # runtime, so only the leading zeros of a value below one are dropped.
    digits = f"{integer_text}{fraction}".lstrip("0")
    exponent = int(exponent_text or 0) - len(fraction) + len(digits) - 1
    mantissa = f"{digits[0]}.{digits[1:]}" if len(digits) > 1 else digits
    return f"{'-' if value < 0 else ''}{mantissa}e{exponent}"


def _canonical_string(value: str) -> str:
    parts = ['"']
    for character in value:
        escape = _SHORT_ESCAPES.get(character)
        if escape is not None:
            parts.append(escape)
            continue
        code_point = ord(character)
        if code_point < 0x20:
            parts.append(f"\\u{code_point:04x}")
            continue
        if 0xD800 <= code_point <= 0xDFFF:
            raise EvidenceSealError("A sealable string must not contain an unpaired surrogate.")
        parts.append(character)
    parts.append('"')
    return "".join(parts)


__all__ = [
    "EVIDENCE_SEAL_ALGORITHM",
    "EVIDENCE_SEAL_SCHEMA_VERSION",
    "EvidenceSealError",
    "canonical_seal_text",
    "seal_sha256",
]

File: studio/frontend/src/evidenceSeal.ts
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial license available
// © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
// © Code 2020–2026 Miroslav Šotek. All rights reserved.
// ORCID: 0009-0009-3560-0851
// Contact: www.anulum.li | protoscience@anulum.li
// SC-NeuroCore — Studio cross-runtime evidence seal

/**
 * Browser half of the Studio evidence seal.
 *
 * `sc_neurocore/studio/evidence_seal.py` is the other half. Both encode a
 * value — not a runtime's rendering of it — so a payload the server sealed and
 * the browser handed back seals to the same digest. `JSON.stringify` alone does
 * not: it writes `1` where Python writes `1.0` and `1e-7` where Python writes
 * `1e-07`, which is why a recorded server seal could not be rechecked against
 * anything that had passed through this runtime.
 *
 * The rules, identical on both sides: object keys ordered by code point; no
 * insignificant whitespace; an integral number written as an integer and any
 * other written in scientific form from its shortest round-tripping digits;
 * minimal RFC 8259 string escapes; and a refusal — never a silent change — for
 * a non-finite number or an unpaired surrogate. The Python half additionally
 * refuses an integer no double holds exactly; this runtime has no such value to
 * refuse, because every number here is already a double.
 */

/** Contract version of the canonical form. */
export const EVIDENCE_SEAL_SCHEMA_VERSION = "studio.evidence-seal.v1" as const;

/** Digest algorithm named in every receipt that carries a seal. */
export const EVIDENCE_SEAL_ALGORITHM = "sha256" as const;

/** Raised when a value cannot be sealed identically in both runtimes. */
export class EvidenceSealError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "EvidenceSealError";
  }
}

const SHORT_ESCAPES: Record<string, string> = {
  "\b": "\\b",
  "\t": "\\t",
  "\n": "\\n",
  "\f": "\\f",
  "\r": "\\r",
  '"': '\\"',
  "\\": "\\\\",
};

/**
 * Return the canonical JSON text used as the seal input.
 *
 * @throws EvidenceSealError when the value would not survive a round trip
 * between the two runtimes unchanged.
 */
export function canonicalSealText(value: unknown): string {
  const parts: string[] = [];
  encode(value, parts);
  return parts.join("");
}

/**
 * Return the SHA-256 hex digest of the canonical form of `value`.
 *
 * Requires Web Crypto, which the Studio frontend already depends on for cart
 * digests.
 */
export async function sealSha256(value: unknown): Promise<string> {
  const data = new TextEncoder().encode(canonicalSealText(value));
  if (typeof globalThis.crypto === "undefined" || !globalThis.crypto.subtle) {
    throw new EvidenceSealError("Web Crypto SHA-256 is required for evidence seals");
  }
  const digest = await globalThis.crypto.subtle.digest("SHA-256", data);
  return Array.from(new Uint8Array(digest))
    .map((byte) => byte.toString(16).padStart(2, "0"))
    .join("");
}

function encode(value: unknown, parts: string[]): void {
  if (value === null) {
    parts.push("null");
    return;
  }
  const kind = typeof value;
  if (kind === "boolean") {
    parts.push(value === true ? "true" : "false");
    return;
  }
  if (kind === "number") {
    parts.push(canonicalNumber(value as number));
    return;
  }
  if (kind === "string") {
    parts.push(canonicalString(value as string));
    return;
  }
  if (Array.isArray(value)) {
    parts.push("[");
    value.forEach((item, index) => {
      if (index) {
        parts.push(",");
      }
      encode(item, parts);
    });
    parts.push("]");
    return;
  }
  if (kind === "object") {
    encodeObject(value as Record<string, unknown>, parts);
    return;
  }
  throw new EvidenceSealError(`${kind} is not a sealable JSON value.`);
}

function encodeObject(value: Record<string, unknown>, parts: string[]): void {
  const keys = Object.keys(value).sort(compareCodePoints);
  parts.push("{");
  keys.forEach((key, index) => {
    if (index) {
      parts.push(",");
    }
    parts.push(canonicalString(key));
    parts.push(":");
    encode(value[key], parts);
  });
  parts.push("}");
}

/**
 * Order strings by code point.
 *
 * The default sort compares UTF-16 code units, which disagrees with Python
 * above the basic multilingual plane; comparing code points makes both
 * runtimes agree.
 */
function compareCodePoints(left: string, right: string): number {
  const leftPoints = Array.from(left, (character) => character.codePointAt(0) ?? 0);
  const rightPoints = Array.from(right, (character) => character.codePointAt(0) ?? 0);
  const shared = Math.min(leftPoints.length, rightPoints.length);
  for (let index = 0; index < shared; index += 1) {
    if (leftPoints[index] !== rightPoints[index]) {
      return leftPoints[index] < rightPoints[index] ? -1 : 1;
    }
  }
  return leftPoints.length - rightPoints.length;
}

function canonicalNumber(value: number): string {
  if (!Number.isFinite(value)) {
    throw new EvidenceSealError(`${String(value)} is not a sealable JSON number.`);
  }
  if (Number.isInteger(value)) {
    // BigInt renders an integral double exactly; toString(10) would switch to
    // exponent notation above 1e21 and disagree with the Python half.
    return BigInt(value).toString(10);
  }
  return canonicalFraction(value);
}

function canonicalFraction(value: number): string {
  const negative = value < 0;
  const text = Math.abs(value).toString(10);
  const matched = /^(\d+)(?:\.(\d+))?(?:[eE]([+-]?\d+))?$/.exec(text);
  if (matched === null) {
    throw new EvidenceSealError(`${text} has no canonical numeric form.`);
  }
  const fraction = matched[2] ?? "";
  // The shortest round-tripping form never ends in a redundant zero, in either
  // runtime, so only the leading zeros of a value below one are dropped.
  const digits = `${matched[1]}${fraction}`.replace(/^0+(?=\d)/, "");
  const exponent = Number(matched[3] ?? "0") - fraction.length + digits.length - 1;
  const mantissa = digits.length > 1 ? `${digits[0]}.${digits.slice(1)}` : digits;
  return `${negative ? "-" : ""}${mantissa}e${exponent}`;
}

function canonicalString(value: string): string {
  const parts = ['"'];
  for (const character of value) {
    const escape = SHORT_ESCAPES[character];
    if (escape !== undefined) {
      parts.push(escape);
      continue;
    }
    const codePoint = character.codePointAt(0) ?? 0;
    if (codePoint < 0x20) {
      parts.push(`\\u${codePoint.toString(16).padStart(4, "0")}`);
      continue;
    }
    if (codePoint >= 0xd800 && codePoint <= 0xdfff) {
      throw new EvidenceSealError(
        "A sealable string must not contain an unpaired surrogate.",
      );
    }
    parts.push(character);
  }
  parts.push('"');
  return parts.join("");
}
